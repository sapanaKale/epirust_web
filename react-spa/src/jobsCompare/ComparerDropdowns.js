/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useState} from "react";
import JobsDropdown from "./JobsDropdown";
import PropTypes from "prop-types";

export default function ComparerDropdowns({jobs, onCompare}) {
  const [showError, updateShowError] = useState(false);

  const handleSubmit = (e) => {
    e.preventDefault();
    const formsData = new FormData(e.target);
    const selectedJobs = Object.fromEntries(formsData.entries())
    if (selectedJobs['job1'] === selectedJobs['job2']) {
      updateShowError(true)
    } else {
      onCompare(selectedJobs)
    }
  }

  return (
    <div className='comparer-dropdowns container'>
      <form className='row' onSubmit={handleSubmit}>
        <div className="col-5">
          <JobsDropdown jobs={jobs} label='Job 1'/>
        </div>
        <div className="col-5">
          <JobsDropdown jobs={jobs} label='Job 2'/>
        </div>
        <div className="col-2 margin-top-auto">
          <button className="btn btn-primary">Compare</button>
        </div>
        {showError && <div className="error-message alert-danger">Can't compare same jobs!</div>}
      </form>
    </div>
  );
}

ComparerDropdowns.propTypes = {
  jobs: PropTypes.array.isRequired,
  onCompare: PropTypes.func
}