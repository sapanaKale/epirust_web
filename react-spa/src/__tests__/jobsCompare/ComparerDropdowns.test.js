/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import {render} from '@testing-library/react'
import ComparerDropdowns from "../../jobsCompare/ComparerDropdowns";
import userEvent from "@testing-library/user-event";

describe('Comparer Dropdowns', () => {
  const testJobs = [{_id: 'id1'}, {_id: 'id2'}];

  it('should render inputs to select jobs to compare', () => {
    const {asFragment} = render(<ComparerDropdowns jobs={testJobs}/>);
    expect(asFragment()).toMatchSnapshot()
  });

  it('should return selected jobs when compare button clicked', () => {
    const mockOnCompare = jest.fn()
    const {getByRole, getByLabelText} = render(<ComparerDropdowns jobs={testJobs} onCompare={mockOnCompare}/>);

    userEvent.selectOptions(getByLabelText('Job 1'), ['id1']);
    userEvent.selectOptions(getByLabelText('Job 2'), ['id2']);
    userEvent.click(getByRole('button', /Compare/));

    expect(mockOnCompare).toHaveBeenCalledTimes(1)
    expect(mockOnCompare).toHaveBeenCalledWith({job1: 'id1', job2: 'id2'})
  });

  it('should show error message if both the selected jobs are same on compare click', () => {
    const {queryByText, getByRole, getByLabelText} = render(<ComparerDropdowns jobs={testJobs}/>);

    userEvent.selectOptions(getByLabelText('Job 1'), ['id1']);
    userEvent.selectOptions(getByLabelText('Job 2'), ['id1']);
    userEvent.click(getByRole('button', /Compare/));

    expect(queryByText("Can't compare same jobs!")).toBeInTheDocument()
  });
});

