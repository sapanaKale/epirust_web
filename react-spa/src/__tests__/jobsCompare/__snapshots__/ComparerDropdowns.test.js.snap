// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comparer Dropdowns should render inputs to select jobs to compare 1`] = `
<DocumentFragment>
  <div
    class="comparer-dropdowns container"
  >
    <form
      class="row"
    >
      <div
        class="col-5"
      >
        <label
          for="job1"
        >
          Job 1
        </label>
        <select
          class="form-control"
          id="job1"
          name="job1"
          required=""
        >
          <option
            disabled=""
            value=""
          >
             -- select a job -- 
          </option>
          <option>
            id1
          </option>
          <option>
            id2
          </option>
        </select>
      </div>
      <div
        class="col-5"
      >
        <label
          for="job2"
        >
          Job 2
        </label>
        <select
          class="form-control"
          id="job2"
          name="job2"
          required=""
        >
          <option
            disabled=""
            value=""
          >
             -- select a job -- 
          </option>
          <option>
            id1
          </option>
          <option>
            id2
          </option>
        </select>
      </div>
      <div
        class="col-2 margin-top-auto"
      >
        <button
          class="btn btn-primary"
        >
          Compare
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
